pinot-segment-local/src/test/java/org/apache/pinot/segment/local/segment/index/loader/ForwardIndexHandlerTest.java

```diff
@@ -0,0 +1,374 @@
+        FieldConfig.CompressionCodec.ZSTANDARD, null);
+    fieldConfigs.add(newConfig);
+    tableConfig =
+        new TableConfigBuilder(TableType.OFFLINE).setTableName(TABLE_NAME).setNoDictionaryColumns(_noDictionaryColumns)
+            .setFieldConfigList(fieldConfigs).build();
+    tableConfig.setFieldConfigList(fieldConfigs);
+
+    indexLoadingConfig = new IndexLoadingConfig(null, tableConfig);
+    indexLoadingConfig.getTextIndexColumns().add(config.getName());
+    indexLoadingConfig.getInvertedIndexColumns().add(config.getName());
+    fwdIndexHandler = new ForwardIndexHandler(_existingSegmentMetadata, indexLoadingConfig);
+    operationMap = fwdIndexHandler.computeOperation(_writer);
+    assertEquals(operationMap.size(), 1);
+    assertEquals(operationMap.get(config.getName()), ForwardIndexHandler.Operation.CHANGE_RAW_COMPRESSION_TYPE);
+  }
+
+  @Test
+  public void testRewriteRawForwardIndex()
+      throws Exception {
+    List<String> allCompressionTypes = Arrays.asList("PASS_THROUGH", "ZSTANDARD", "LZ4", "SNAPPY");
+
+    for (int i = 0; i < _noDictionaryColumns.size(); i++) {
+
+      // For every noDictionaryColumn, change the compressionType to all available types, one by one.
+      for (String compressionType : allCompressionTypes) {
+        List<FieldConfig> fieldConfigs = new ArrayList<>(_tableConfig.getFieldConfigList());
+        FieldConfig config = fieldConfigs.remove(i);
+        String columnName = config.getName();
+        FieldConfig.CompressionCodec newCompressionType = FieldConfig.CompressionCodec.valueOf(compressionType);
+        if (config.getCompressionCodec().equals(newCompressionType)) {
+          // Compression types match. No need to invoke forward index handler.
+          continue;
+        }
+
+        FieldConfig newConfig =
+            new FieldConfig(columnName, FieldConfig.EncodingType.RAW, Collections.emptyList(), newCompressionType,
+                null);
+        fieldConfigs.add(newConfig);
+
+        TableConfig tableConfig = new TableConfigBuilder(TableType.OFFLINE).setTableName(TABLE_NAME)
+            .setNoDictionaryColumns(_noDictionaryColumns).setFieldConfigList(fieldConfigs).build();
+        tableConfig.setFieldConfigList(fieldConfigs);
+
+        IndexLoadingConfig indexLoadingConfig = new IndexLoadingConfig(null, tableConfig);
+        IndexCreatorProvider indexCreatorProvider = IndexingOverrides.getIndexCreatorProvider();
+        ForwardIndexHandler fwdIndexHandler =
+            new ForwardIndexHandler(_existingSegmentMetadata, indexLoadingConfig);
+        fwdIndexHandler.updateIndices(_writer, indexCreatorProvider);
+
+        ColumnMetadata columnMetadata = _existingSegmentMetadata.getColumnMetadataFor(columnName);
+
+        validateForwardIndex(columnMetadata, newCompressionType);
+      }
+    }
+  }
+
+  private void validateForwardIndex(ColumnMetadata columnMetadata, FieldConfig.CompressionCodec expectedCompressionType)
```

Comment:
Possible / need to check if the offsets / index_map file is correct ? Can we assert for start offset and length of forward index buffer

Reply:
As discussed offline, added a check to see if forward index offsets are present in index_map. As we anyways check compressionType and validate raw values, this should be sufficient.